### PR TITLE
Make /state endpoint asynchronous

### PR DESCRIPTION
**Pull request by shamsimam**

### CODE REVIEW

File: waiter/src/waiter/handler.clj [hunk 1]
```diff
@@ -493,25 +492,34 @@
...
   (try
-    (let [timeout-ms 30000
+    (let [embed-param (-> request
+                          ring-params/params-request
+                          :params
+                          (get "embed"))
+          embed-pred (cond
+                       (string? embed-param) #{embed-param}
+                       (sequential? embed-param) (set embed-param)
+                       :else (constantly true))
```
Comment (sradack) [inline waiter/src/waiter/handler.clj]: How do I just get state?

File: waiter/src/waiter/handler.clj [hunk 2]
```diff
@@ -493,25 +492,34 @@
...
+                                       (async/>!! scheduler-chan {:response-chan response-chan})
+                                       (async/alt!!
+                                         response-chan ([state] state)
+                                         (async/timeout timeout-ms) ([_] {:message "Query for scheduler state timed out"})
+                                         :priority true)))]
+      (-> (cond-> current-state
+                  (embed-pred "kv-store") (assoc :kv-store(kv/state kv-store))
+                  (embed-pred "leader") (assoc :leader (leader?-fn))
+                  (embed-pred "router-metrics-state") (assoc :router-metrics-state (router-metrics-state-fn))
+                  (embed-pred "scheduler") (assoc :scheduler (retrieve-scheduler-state))
```
Comment (sradack) [inline waiter/src/waiter/handler.clj]: Please update the tests to specify the embeds they need.

File: waiter/src/waiter/handler.clj [hunk 3]
```diff
@@ -493,25 +492,34 @@
 
 (defn get-router-state
   "Outputs the state of the router as json."
-  [state-chan scheduler-chan router-metrics-state-fn kv-store leader?-fn scheduler request]
+  [state-chan scheduler-chan router-metrics-state-fn kv-store leader?-fn request]
   (try
-    (let [timeout-ms 30000
+    (let [embed-param (-> request
```
Comment (sradack) [inline waiter/src/waiter/handler.clj]: Let's make this method async so we don't block a thread.

File: waiter/src/waiter/handler.clj [hunk 4]
```diff
@@ -493,25 +492,34 @@
...
-                              response-chan ([state] state)
-                              (async/timeout timeout-ms) ([_] {:message "Query for scheduler state timed out"})
-                              :priority true))]
-      (-> current-state
-          (assoc :leader (leader?-fn)
-                 :kv-store (kv/state kv-store)
-                 :router-metrics-state (router-metrics-state-fn)
-                 :scheduler scheduler-state
-                 :statsd (statsd/state))
+          retrieve-scheduler-state (fn retrieve-scheduler-state-fn []
```
Comment (sradack) [inline waiter/src/waiter/handler.clj]: If async, this can't be in a function.

File: waiter/src/waiter/handler.clj (hunk 1 above)
Comment (shamsimam) [reply waiter/src/waiter/handler.clj]: state is always included by default. So passing `embed=state` will achieve that.

Comment (sradack) [reply waiter/src/waiter/handler.clj]: While `embed=state` works given the implementation, it's an awkward API. We don't "embed" state within state.  

We have an opportunity to clean things up, making `/state`, by default, return just state, and embed options add to the output.

Another opportunity is to create dedicated endpoints for the specific states, e.g.
`/state/scheduler`
`/state/statsd`
etc.

File: waiter/src/waiter/handler.clj (hunk 2 above)
Comment (shamsimam) [reply waiter/src/waiter/handler.clj]: Done.

File: waiter/src/waiter/handler.clj (hunk 4 above)
Comment (shamsimam) [reply waiter/src/waiter/handler.clj]: Converted to async.

File: waiter/src/waiter/handler.clj (hunk 3 above)
Comment (shamsimam) [reply waiter/src/waiter/handler.clj]: Converted to async.

**Comment by shamsimam** [comment]
RTM

**Comment by shamsimam** [comment]
RTM, the scope of the PR has changed to make `/state` handler become asynchronous.

**Pull request merged**